chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,338 @@
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{error::ConnectTransactionError, TransactionSource};
+
+use common::{
+    chain::{Block, PoolId},
+    primitives::{Amount, Id},
+    Uint256,
+};
+use pos_accounting::{AccountingBlockRewardUndo, PoSAccountingOperations, PoSAccountingView};
+
+use super::accounting_delta_adapter::PoSAccountingDeltaAdapter;
+
+/// Distribute reward among the staker and delegators
+pub fn distribute_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_reward: Amount,
+) -> Result<AccountingBlockRewardUndo, ConnectTransactionError> {
+    let tx_source = TransactionSource::Chain(block_id);
+
+    let pool_data = accounting_adapter
+        .accounting_delta()
+        .get_pool_data(pool_id)?
+        .ok_or(ConnectTransactionError::PoolDataNotFound(pool_id))?;
+
+    let total_staker_reward = (total_reward - pool_data.cost_per_block())
+        .and_then(|v| (v * pool_data.margin_ratio_per_thousand().into()).and_then(|v| v / 1000))
+        .and_then(|v| v + pool_data.cost_per_block())
+        .ok_or(ConnectTransactionError::StakerRewardCalculationFailed(
+            block_id,
+        ))?;
+    let total_delegators_reward = (total_reward - total_staker_reward).ok_or(
+        ConnectTransactionError::DelegatorRewardCalculationFailed(block_id),
+    )?;
+    let total_delegators_reward = Uint256::from_amount(total_delegators_reward);
+
+    let increase_balance_undo = accounting_adapter
+        .operations(tx_source)
+        .increase_pool_balance(pool_id, total_staker_reward)?;
+
+    let delegation_undos = accounting_adapter
+        .accounting_delta()
+        .get_pool_delegations_shares(pool_id)?
+        .map(|delegation_shares| {
+            let total_delegators_balance =
+                delegation_shares.values().try_fold(Amount::ZERO, |acc, v| {
+                    (acc + *v).ok_or(ConnectTransactionError::DelegatorsRewardSumFailed(block_id))
+                })?;
+            let total_delegators_balance = Uint256::from_amount(total_delegators_balance);
+
+            delegation_shares
+                .iter()
+                .map(|(delegation_id, balance)| {
+                    let balance = Uint256::from_amount(*balance);
+                    let reward = total_delegators_reward * balance / total_delegators_balance;
```

Comment:
I have a feeling you're doing this and thinking that you're "covered with Uint256 because the precision is twice the multiplicands"... but you're not doing this right. In this line, first, `balance / total_delegators_balance` will happen first, and then the result will be multiplied by `total_delegators_reward`. If this is what you really intended, then I gotta ask why you converted to Uint256 in the first place.

Reply:
Why division should happen first? 
I thought that operations are left to right as per: https://doc.rust-lang.org/reference/expressions.html#expression-precedence

Reply:
Division always happens first in arithmetic. I'll double check to see why rust is claiming to do it like this.

Reply:
`println!("{}", 3 * 3 / 2);` gives me 4 not 3

Reply:
I'll add parentheses anyway to avoid confusion in the future

Reply:
That's better. 

Weird. I could swear it's not like that in Mathemtica and C++, but now I have to recheck my understanding of this.